### PR TITLE
JAMES-2586 - Bump Jooq 3.19.13 -> 3.19.15 & resolve hanging issue

### PR DESCRIPTION
**Pull request by vttranlina**

### CODE REVIEW

File: backends-common/postgres/src/main/java/org/apache/james/backends/postgres/utils/PostgresExecutor.java [hunk 1]
```diff
@@ -145,8 +143,6 @@ public Flux<Record> executeDeleteAndReturnList(Function<DSLContext, DeleteResult
                     .flatMapMany(queryFunction)
                     .timeout(postgresConfiguration.getJooqReactiveTimeout())
                     .doOnError(TimeoutException.class, e -> LOGGER.error(JOOQ_TIMEOUT_ERROR_LOG, e))
-                    .collectList()
```
Comment (vttranlina) [inline backends-common/postgres/src/main/java/org/apache/james/backends/postgres/utils/PostgresExecutor.java:148]: The hanging issue occurs when connection-A is processing a query, and another where of the code tries to use connection-A for a different query. 

I guess this issue might be resolved when we using a connection pool (implemented)
Wait to CI result for confirmation

**Review by quantranhong1999 — APPROVED** [review]

**Comment by chibenwa** [comment]
WOOT!

**Review by chibenwa — APPROVED** [review]

**Comment by chibenwa** [comment]
Did we just kill it?

Can I open champagne tonight?

**Comment by vttranlina** [comment]
Although using a connection pool is a "application layer" solution that can fix this issue, but IMO, it does not solve the root cause. 
The connection pool size must be "large enough" to avoid errors.

This can be better understood through the following example:
Give: Connection pool size = 3
When reactive chain logic:
```
getUser  
.flatMap(user -> getMailbox)  
.flatMap(mailbox -> getMessages)  
.flatMap(message -> updateQuota)
```

You can see in the chain that we need 4 connections available at the same time.

4 > 3 => We will get an exception.

**Comment by chibenwa** [comment]
Any way we might get something prod worthy with a big enough connection pool I bet.

Good news still, though underlying issue might need investigations.

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
Hmmmm... related? ^^'

https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-2514/3/testReport/org.apache.james.mailbox.postgres.mail.task/PostgresRecomputeCurrentQuotasServiceTest/recomputeCurrentQuotasShouldUpdateContextWhenIncorrectQuotas/

**Comment by chibenwa** [comment]
(Indeed)

image

ChatGPT chat if relevant: this *might* be an actuall PG limitation we hit...

```
Is a SQL connection usable to run other queries if I am paging results from the database on a select query?
```

Whether a SQL connection can be used to run other queries while paging results from a database depends on the specifics of how the database and the connection are configured, as well as the SQL client library you're using. Here are the key considerations:

### 1. **Cursor Behavior**
   - **Forward-Only Cursors**: If you're using a forward-only cursor for paging (common in libraries such as JDBC or Python's `psycopg2`), the connection is typically locked for that query, and running another query might not be allowed until the current one completes or the cursor is closed.
   - **Scrollable Cursors**: Some databases support scrollable cursors, which might allow other queries on the same connection depending on the database's implementation.

### 2. **Database Locking and Transaction Scope**
   - If the paging query is running within a transaction, the database may hold locks that prevent concurrent queries on the same connection.
   - Running multiple queries simultaneously on a single connection could also violate transactional guarantees.

### 3. **Database Support for Concurrent Queries**
   - Some databases (e.g., PostgreSQL, MySQL) support only one active query per connection at a time. Running another query will result in an error unless the first query is completed or the results are fully fetched.
   - Databases like SQL Server and Oracle allow multiple active result sets (MARS) or similar mechanisms, enabling simultaneous queries under certain conditions.

### 4. **Library-Specific Considerations**
   - Some client libraries (like Python's `asyncpg` for PostgreSQL) allow multiple queries to be queued asynchronously on a single connection.
   - Libraries like `sqlalchemy` in Python often manage connection pooling, enabling multiple queries across multiple connections, even if one connection is busy.

### Best Practices for Paging
- **Use Connection Pooling**: Tools like HikariCP (Java) or SQLAlchemy’s connection pool ensure each query can use its own connection, avoiding contention.
- **Fetch Small Chunks**: Retrieve only the required number of rows for each page to minimize connection usage.
- **Close Cursors**: Ensure cursors are closed promptly after fetching the results.

### Conclusion
If you're paging results from a database on a select query, it's generally safer to assume that the connection is not available for other queries until the operation completes. Using connection pooling or multiple connections can mitigate this limitation effectively.

```
Does postgresql support scrollable cursors?

Are scrollable cursors implemented with R2DC library?
```

### PostgreSQL and Scrollable Cursors

**1. PostgreSQL Native Support:**
PostgreSQL does support scrollable cursors. A scrollable cursor allows you to fetch rows in any order (e.g., forward, backward, relative, absolute positioning). This is particularly useful for applications that need random access to query results.

- To create a scrollable cursor in PostgreSQL, you use the SQL `DECLARE` statement with the `SCROLL` keyword:

```sql
DECLARE my_cursor SCROLL CURSOR FOR SELECT * FROM my_table;
```

- Once the cursor is declared, you can use `FETCH` commands like `FETCH NEXT`, `FETCH PRIOR`, or `FETCH ABSOLUTE` to navigate through the result set.

- Scrollable cursors are limited to explicit transactions. That means you need to wrap them within a `BEGIN`/`COMMIT` or `BEGIN`/`ROLLBACK` block.

---

**2. R2DBC and Scrollable Cursors:**
[R2DBC](https://r2dbc.io/) (Reactive Relational Database Connectivity) is designed for reactive programming and focuses on streaming data asynchronously. 

- R2DBC itself does not directly expose scrollable cursors in the way traditional JDBC does. Instead, it provides reactive streams for processing results row-by-row in a non-blocking manner.
  
- While PostgreSQL supports scrollable cursors natively, the R2DBC PostgreSQL driver does not provide a direct interface for them. You would need to implement a workaround using PostgreSQL's native SQL commands for cursor operations.

---

### Summary

- **PostgreSQL:** Fully supports scrollable cursors with SQL commands.
- **R2DBC:** Does not natively implement scrollable cursors. However, its reactive streams can achieve similar functionality by processing and re-querying data as needed for reactive workflows.

If scrollable cursors are a strict requirement for your application, traditional JDBC might be a better fit than R2DBC.

**Comment by chibenwa** [comment]
@vttranlina does JOOQ offers some way to do lazy fetching in a reactive fashion and does it help?

Discussed with @yadd:

```
execute(select().from(MESSAGES))
```

Is syntactically equivalent to:

```
// Execute as many time as needed
execute(select().from(MESSAGES))
    .sortOn(UID)
    .where(UID > highestUID so far)
    .limit(pageSize)
```

IE given a mailbox with 2500 messages we execute the select 3 times.

WHat it effectively does in implement paging within the application.

We could:
 - .collectList when working with users, user mailboxes and maybe even mailboxes. COunt of entity is low.
 - However maybe messages would be the only place where such kind of fetching really make sense ?

**Comment by vttranlina** [comment]
> Hmmmm... related? ^^'
> 
> https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-2514/3/testReport/org.apache.james.mailbox.postgres.mail.task/PostgresRecomputeCurrentQuotasServiceTest/recomputeCurrentQuotasShouldUpdateContextWhenIncorrectQuotas/

yes, that I posted at https://github.com/apache/james-project/pull/2514#issuecomment-2490210967
The test using `PoolSize.SMALL(1, 2)`, that is reason

**Comment by hungphan227** [comment]
I have an idea: If a connection can be used safely by different entities in the same thread, we can share connections per thread -> reduce number of connections.

**Comment by vttranlina** [comment]
>  If a connection can be used safely by different entities in the same thread

"entities", what your mean?

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxMessageDAO.java [hunk 2]
```diff
@@ -117,78 +116,107 @@ public PostgresMailboxMessageDAO create(Optional<Domain> domain) {
...
+    private Mono<List<MessageUid>> listUnseenBatch(PostgresMailboxId mailboxId, int offset, int batchSize) {
+        return postgresExecutor.executeRows(dslContext -> Flux.from(dslContext.select(MESSAGE_UID)
+                .from(TABLE_NAME)
+                .where(MAILBOX_ID.eq((mailboxId.asUuid())))
+                .and(IS_SEEN.eq(false))
+                .orderBy(DEFAULT_SORT_ORDER_BY)
+                .limit(batchSize)
+                .offset(offset)))
+            .map(RECORD_TO_MESSAGE_UID_FUNCTION)
+            .collectList()
```
Comment (vttranlina) [inline mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxMessageDAO.java:151]: Note: we still need `collectionList` in small block. 
If not, the issue https://github.com/apache/james-project/pull/2514#issuecomment-2490210967 still happend.

TLDR: if we don't want to use `collectionList` 
the batchSize must be < `32` when reactor configuration by default (IMO, 32 is quiet small for batch size)

image

File: backends-common/postgres/src/main/java/org/apache/james/backends/postgres/PostgresCommons.java [hunk 3]
```diff
@@ -93,4 +93,8 @@ public static <T> Field<T> tableField(Table<Record> table, Field<T> field) {
 
     public static final int IN_CLAUSE_MAX_SIZE = 32;
 
+    public static final int BATCH_SIZE = 100;
```
Comment (chibenwa) [inline backends-common/postgres/src/main/java/org/apache/james/backends/postgres/PostgresCommons.java:96]: Make this configurable?

File: backends-common/postgres/src/main/java/org/apache/james/backends/postgres/utils/PostgresExecutor.java [hunk 4]
```diff
@@ -125,16 +126,26 @@ public Mono<Void> executeVoid(Function<DSLContext, Mono<?>> queryFunction) {
...
-                        .filter(preparedStatementConflictException())),
+                connection -> {
+                    Flux<Record> recordFlux = dslContext(connection)
+                        .flatMapMany(queryFunction)
+                        .timeout(postgresConfiguration.getJooqReactiveTimeout())
+                        .doOnError(TimeoutException.class, e -> LOGGER.error(JOOQ_TIMEOUT_ERROR_LOG, e))
+                        .retryWhen(Retry.backoff(MAX_RETRY_ATTEMPTS, MIN_BACKOFF)
+                            .filter(preparedStatementConflictException()));
+
+                    if (isEagerFetch) {
```
Comment (chibenwa) [inline backends-common/postgres/src/main/java/org/apache/james/backends/postgres/utils/PostgresExecutor.java:143]: Link an issue in a comment to explain this.

File: backends-common/postgres/src/main/java/org/apache/james/backends/postgres/utils/PostgresExecutor.java [hunk 5]
```diff
@@ -125,16 +126,26 @@ public Mono<Void> executeVoid(Function<DSLContext, Mono<?>> queryFunction) {
     }
 
     public Flux<Record> executeRows(Function<DSLContext, Flux<Record>> queryFunction) {
+        return executeRows(queryFunction, !EAGER_FETCH);
+    }
+
+    public Flux<Record> executeRows(Function<DSLContext, Flux<Record>> queryFunction, boolean isEagerFetch) {
```
Comment (chibenwa) [inline backends-common/postgres/src/main/java/org/apache/james/backends/postgres/utils/PostgresExecutor.java:132]: Document in a comment what `isEagerFetch` is doing, why it is needed, when it is Ok to use, when it is not Ok to use, and which are the alternative I can use.

**Review by chibenwa — APPROVED** [review]
Performance test this changeset?

File: server/apps/postgres-app/sample-configuration/postgres.properties [hunk 6]
```diff
@@ -42,4 +42,7 @@ by-pass-rls.pool.max.size=10
 ssl.mode=allow
 
 ## Duration. Optional, defaults to 10 second. jOOQ reactive timeout when executing Postgres query. This setting prevent jooq reactive bug from causing hanging issue.
-#jooq.reactive.timeout=10second
\ No newline at end of file
+#jooq.reactive.timeout=10second
+
+# Integer. Optional, defaults to 100. In case of large data, this argument specifies the maximum number of rows to return in a single batch set when executing query.
+#jooq.reactive.timeout=100
```
Comment (chibenwa) [inline server/apps/postgres-app/sample-configuration/postgres.properties:48]: Oups copy n paster mistake?

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxMessageDAO.java [hunk 7]
```diff
@@ -117,78 +115,110 @@ public PostgresMailboxMessageDAO create(Optional<Domain> domain) {
...
-                IS_SEEN.eq(false), Limit.unlimited(), dslContext)))
-            .map(RECORD_TO_MESSAGE_UID_FUNCTION);
+        return Flux.range(0, OFFSET_UNLIMITED)
+            .concatMap(offsetIndex -> listUnseenBatch(mailboxId, offsetIndex * queryBatchSize, queryBatchSize))
+            .takeUntil(List::isEmpty)
+            .flatMapIterable(Function.identity());
+    }
+
+    private Mono<List<MessageUid>> listUnseenBatch(PostgresMailboxId mailboxId, int offset, int batchSize) {
+        return postgresExecutor.executeRows(dslContext -> Flux.from(dslContext.select(MESSAGE_UID)
```
[the remark this answers is not in the document]
Comment (vttranlina) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxMessageDAO.java:145]: That is also my concern.
 `memory resource` (only uid) is not only that we must care. when the number of row is large, the batch help we early release connection to pool, for next round.

**Comment by vttranlina** [comment]
> Let's look for bugs: can we mark 1000 messages as deleted and then expunge and ensure we did delete then all?
> Also keep in mind 'application batching' needs sort order in order to be deterministic.

👍 

> Is there other use cases where lack of snapshot reads really hurt?

> Can we think of tasks iterating messages and deleting them?

Do you mean that during a batch query, the data might change, causing the query result to become outdated?  
And potentially, the result of the next batch pagination query could be incorrect (especially in the case of deletions, as it changes the total size)?

**Comment by vttranlina** [comment]
Fixup `.collectList` first for CI checking

I'm processing 
> Let's look for bugs: can we mark 1000 messages as deleted and then expunge and ensure we did delete then all?

**Review by quantranhong1999 — COMMENTED** [review]
Read it.

File: server/blob/blob-postgres/src/main/java/org/apache/james/blob/postgres/PostgresBlobStoreDAO.java [hunk 8]
```diff
@@ -151,9 +165,20 @@ public Flux<BucketName> listBuckets() {
...
+    }
+
+    private Mono<List<BlobId>> listBlobsBatch(BucketName bucketName,  int offset, int batchSize) {
         return postgresExecutor.executeRows(dsl -> Flux.from(dsl.select(BLOB_ID)
                 .from(TABLE_NAME)
-                .where(BUCKET_NAME.eq(bucketName.asString()))))
-            .map(record -> blobIdFactory.parse(record.get(BLOB_ID)));
+                .where(BUCKET_NAME.eq(bucketName.asString()))
+                .limit(batchSize)
+                .offset(offset)))
```
[the remark this answers is not in the document]
Comment (vttranlina) [reply server/blob/blob-postgres/src/main/java/org/apache/james/blob/postgres/PostgresBlobStoreDAO.java]: > We are missing ordering to ensure the stability of the transformation.

updated

> Also here we have a massive issue: the use case is to delete as we iterate. This means that, without snapshot reads, the writes would impact the reads and cause entries to be skept.

Indeed, that's correct. However, this is a GC feature. While batch commands may not result in thorough cleanup initially, with repeated GC runs, the data will gradually become more accurate.
What do you think? `collectList` or NOT?

**Comment by vttranlina** [comment]
> Let's look for bugs: can we mark 1000 messages as deleted and then expunge and ensure we did delete then all?

I added the test case `expungeMessageShouldCorrectWhenALotOfMessages`. It always works because `retrieveMessagesMarkedForDeletion` returns collectionList, 
And we use `.collectionList` at DAO layer

**Comment by vttranlina** [comment]
## Performance test result
I noticed that the overall performance did not change significantly.
Some results even showed improvements, but in my opinion, the differences are not very clear.

(likely due to from infrastructor the test environment leading to variations in results across repeated runs of the same test scenario.)

Detail: 
### Gatling result
- with IMAP, 10k user, Duration= 30 minutes 
image

### Metric
- Before
```
postgres_execution{quantile="0.5",} 0.001622015
postgres_execution{quantile="0.75",} 0.0028016630000000002
postgres_execution{quantile="0.95",} 0.021495807000000002
postgres_execution{quantile="0.98",} 0.07864319900000001
postgres_execution{quantile="0.99",} 0.149946367
postgres_execution{quantile="0.999",} 0.599785471
postgres_execution_count 3355670.0
```
- After
```
postgres_execution{quantile="0.5",} 0.0017530870000000002
postgres_execution{quantile="0.75",} 0.0025722870000000004
postgres_execution{quantile="0.95",} 0.005931007
postgres_execution{quantile="0.98",} 0.009895935
postgres_execution{quantile="0.99",} 0.016056319
postgres_execution{quantile="0.999",} 0.057147391000000006
postgres_execution_count 2591175.0
```

### pg_stat_statement
1.  Focus to pagination query

**A. Before**
```sql
select * from message_mailbox where( mailbox_id = cast($1 as uuid) and is_recent = $2) order by message_uid asc
```
- mean_exec_time = 0.36819846626378827
- total_exec_time = 18826.72397700015
- calls = 51132

```sql
select * from message_mailbox where( mailbox_id = cast($1 as uuid) and message_uid >= $2 and message_uid <= $3) order by message_uid asc
```

- mean_exec_time = 0.08792518239753881
- total_exec_time = 2400.621255000017
- calls = 27303
 
**B. After**

```sql
select * from message_mailbox where( mailbox_id = cast($1 as uuid) and is_recent = $2) order by message_uid asc offset $3 rows fetch next $4 rows only
```
- mean_exec_time = 0.12503941170118069
- total_exec_time = 6599.705188999946
- calls = 52781

```sql
select * from message_mailbox where( mailbox_id = cast($1 as uuid) and message_uid >= $2 and message_uid <= $3) order by message_uid asc offset $4 rows fetch next $5 rows only
```

- mean_exec_time = 0.044664966889503485
- total_exec_time = 2455379999974
- calls = 57142


3. Overview
- Before
image

- After
image


// File Result Details: https://github.com/vttranlina/vttranlina.github.io/tree/main/gatling_20241126

**Comment by chibenwa** [comment]
> I noticed that the overall performance did not change significantly.

Almost halth mean time, more than halth for requests relying solely on PGSQL, the results are actually CRAZY good.

This tends to demonstrate a lower contention onto the PG query execution. The metric you provided shows a `x10` drop in tail latency (note that mean latency is unchanged). THis drop in tail latency in the execution of just a few requests fully explains the overall good gains.

**Review by chibenwa — APPROVED** [review]

File: backends-common/postgres/src/main/java/org/apache/james/backends/postgres/PostgresConfiguration.java [hunk 9]
```diff
@@ -59,6 +59,8 @@ public class PostgresConfiguration {
     public static final String SSL_MODE_DEFAULT_VALUE = "allow";
     public static final String JOOQ_REACTIVE_TIMEOUT = "jooq.reactive.timeout";
     public static final Duration JOOQ_REACTIVE_TIMEOUT_DEFAULT_VALUE = Duration.ofSeconds(10);
+    public static final String QUERY_BATCH_SIZE = "query.batch.size";
+    public static final int QUERY_BATCH_SIZE_DEFAULT_VALUE = 5000;
```
Comment (chibenwa) [inline backends-common/postgres/src/main/java/org/apache/james/backends/postgres/PostgresConfiguration.java:63]: Use a system property and default to 5000

File: server/blob/blob-postgres/src/test/java/org/apache/james/blob/postgres/PostgresBlobStoreGCAlgorithmTest.java [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+    @BeforeEach
+    void beforeEach() {
+        blobStore = new PostgresBlobStoreDAO(postgresExtension.getDefaultPostgresExecutor(), new PlainBlobId.Factory(), 1000);
+    }
+
+    @Override
+    public BlobStoreDAO blobStoreDAO() {
+        return blobStore;
+    }
```
[the remark this answers is not in the document]
Comment (vttranlina) [reply server/blob/blob-postgres/src/test/java/org/apache/james/blob/postgres/PostgresBlobStoreGCAlgorithmTest.java:43]: it failed, I tested it 
The reason as you write at https://github.com/apache/james-project/pull/2514#discussion_r1853900136

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxMessageDAO.java [hunk 11]
```diff
@@ -273,31 +270,55 @@ public Mono<Pair<Integer, Integer>> countTotalAndUnseenMessagesByMailboxId(Postg
...
-                .orElse(Flux.from(queryWithoutLimit.apply(dslContext))))
-            .map(record -> Pair.of(fetchStrategy.toMessageBuilder().apply(record), record));
+    private Mono<List<Pair<SimpleMailboxMessage.Builder, Record>>> findMessagesByMailboxIdBatch(PostgresMailboxId mailboxId, MessageMapper.FetchType fetchType, int offset, int batchSize) {
+        PostgresMailboxMessageFetchStrategy fetchStrategy = FETCH_TYPE_TO_FETCH_STRATEGY.apply(fetchType);
+        return postgresExecutor.executeRows(dslContext -> Flux.from(dslContext.select(fetchStrategy.fetchFields())
+                .from(MESSAGES_JOIN_MAILBOX_MESSAGES_CONDITION_STEP)
+                .where(MAILBOX_ID.eq(mailboxId.asUuid()))
+                .orderBy(DEFAULT_SORT_ORDER_BY)
+                .limit(batchSize)
+                .offset(offset)))
```
Comment (chibenwa) [inline mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxMessageDAO.java:291]: Actually we would be WAY more resilient not using an offset. We SHALL reuse the highest UID from the previous batch and resume from it. This would make us reslient to deletion in already paged items.

**Review by chibenwa — CHANGES_REQUESTED** [review]

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxMessageDAO.java (hunk 11 above)
Comment (vttranlina) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxMessageDAO.java:291]: Good idea
Should we also implement this for the Postgres BlobStore DAO?
We can add an additional column, such as a sequence or counter column?

Comment (chibenwa) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxMessageDAO.java:291]: > Should we also implement this for the Postgres BlobStore DAO?

It would rely on alphabetic ordering, it is likely less important.

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxMessageDAO.java (hunk 2 above)
Comment (hungphan227) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxMessageDAO.java]: We remove collectList but still use it in some cases. Really complicated to me.

Comment (vttranlina) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxMessageDAO.java]: https://github.com/apache/james-project/pull/2514#issuecomment-2490348291

File: server/blob/blob-postgres/src/test/java/org/apache/james/blob/postgres/PostgresBlobStoreGCAlgorithmTest.java [hunk 12]
```diff
@@ -50,6 +49,7 @@ public class PostgresBlobStoreGCAlgorithmTest implements BloomFilterGCAlgorithmC
 
     @BeforeEach
     void beforeEach() {
+        System.setProperty("james.postgresql.query.batch.size", "10");
```
Comment (chibenwa) [inline server/blob/blob-postgres/src/test/java/org/apache/james/blob/postgres/PostgresBlobStoreGCAlgorithmTest.java:52]: Do not forget to unset it at the end of the test

**Review by chibenwa — COMMENTED** [review]
Thanks for your patience @vttranlina the resulting work looks awesome.

Maybe we can try to run a performance test just to confirm the current version is on par with the previous one?

**Review by chibenwa — APPROVED** [review]

**Comment by chibenwa** [comment]
Please squash

**Review by Arsnael — APPROVED** [review]

**Review by hungphan227 — APPROVED** [review]

**Comment by vttranlina** [comment]
# Performance test result 

When running performance tests for the latest code, I observed a degradation in some result index compared to the previous run. Suspecting potential system infrastructure instability, I decided to rerun the performance tests for both the "before" and "after" states of the PR.

The detailed results are shown below.

## Gatling
image


## Metrics
- Before
```
# TYPE postgres_execution summary
postgres_execution{quantile="0.5",} 0.0019005430000000002
postgres_execution{quantile="0.75",} 0.0029982710000000003
postgres_execution{quantile="0.95",} 0.008323071000000001
postgres_execution{quantile="0.98",} 0.014811135000000001
postgres_execution{quantile="0.99",} 0.024379391
postgres_execution{quantile="0.999",} 0.071827455
postgres_execution_count 2149596.0
```

- After
```
# TYPE postgres_execution summary
postgres_execution{quantile="0.5",} 0.0015646710000000001
postgres_execution{quantile="0.75",} 0.002260991
postgres_execution{quantile="0.95",} 0.005013503
postgres_execution{quantile="0.98",} 0.007438335000000001
postgres_execution{quantile="0.99",} 0.009961471000000001
postgres_execution{quantile="0.999",} 0.028311551
postgres_execution_count 2244489.0
```

## pg_stat_statement
- Before
image

- After
image


File test result: https://github.com/vttranlina/vttranlina.github.io/tree/main/gatling_20241127
